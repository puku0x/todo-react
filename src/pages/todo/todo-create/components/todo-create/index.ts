export * from './todo-create.component';
