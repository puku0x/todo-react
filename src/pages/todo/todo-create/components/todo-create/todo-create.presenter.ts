import { useMemo } from 'react';
import { useFormik } from 'formik';
import * as Yup from 'yup';

import { TodoCreateDto } from '../../../../../models';

interface FormValues {
  title: string;
}

const toDto = (values: FormValues) => {
  const value: TodoCreateDto = {
    title: values.title,
  };
  return value;
};

export const useTodoCreatePresenter = (arg: {
  onCreate?: (todo: TodoCreateDto) => void;
}) => {
  const { onCreate } = arg;

  const initialValues = useMemo(() => {
    return {
      title: '',
    } as FormValues;
  }, []);

  const validationSchema = useMemo(() => {
    return Yup.object().shape({
      title: Yup.string().required(),
    });
  }, []);

  const formik = useFormik({
    enableReinitialize: true,
    initialValues,
    validateOnMount: true,
    validationSchema,
    onSubmit: (values) => {
      onCreate?.(toDto(values));
    },
  });

  return {
    ...formik,
  } as const;
};
