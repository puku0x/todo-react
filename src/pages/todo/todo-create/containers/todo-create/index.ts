export * from './todo-create.container';
