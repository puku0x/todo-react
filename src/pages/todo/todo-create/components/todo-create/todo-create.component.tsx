import * as React from 'react';
import { NavLink } from 'react-router-dom';

import { TodoCreateDto } from '../../../../../models';
import { useTodoCreatePresenter } from './todo-create.presenter';

type Props = {
  isFetching?: boolean;
  onCreate?: (todo: TodoCreateDto) => void;
};

export const TodoCreate: React.FC<Props> = React.memo((props) => {
  const { isFetching, onCreate } = props;

  const {
    isValid,
    handleBlur,
    handleChange,
    handleSubmit,
  } = useTodoCreatePresenter({ onCreate });

  return (
    <>
      <NavLink to="/todos">Back to list</NavLink>
      <h2>todo-create</h2>
      <form onSubmit={handleSubmit}>
        <p>
          <button
            type="submit"
            disabled={isFetching || !isValid}
            onBlur={handleBlur}
            onChange={handleChange}
          >
            Save
          </button>
        </p>
        <table>
          <tbody>
            <tr>
              <td>title</td>
              <td>
                <input
                  type="text"
                  name="title"
                  onBlur={handleBlur}
                  onChange={handleChange}
                />
              </td>
            </tr>
          </tbody>
        </table>
      </form>
    </>
  );
});
