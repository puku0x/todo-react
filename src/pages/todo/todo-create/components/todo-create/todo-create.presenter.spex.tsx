import { act, renderHook } from '@testing-library/react-hooks';

import { useTodoCreatePresenter } from './todo-create.presenter';

describe('useTodoCreateForm', () => {
  it('should handle submit', async () => {
    const onCreate = jest.fn();
    const { result, waitFor } = renderHook(() =>
      useTodoCreatePresenter({ onCreate })
    );

    await act(async () => {
      const { values, setValues } = result.current;
      setValues({
        ...values,
        title: 'title',
      });
    });

    await act(async () => {
      const { handleSubmit } = result.current;
      handleSubmit();
    });

    await waitFor(() => {
      expect(onCreate).toHaveBeenCalled();
    });
  });
});
