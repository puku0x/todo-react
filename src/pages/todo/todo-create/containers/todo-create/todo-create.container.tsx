import { memo } from 'react';

import { TodoCreate } from '../../components';
import { useTodoCreateFacade } from './todo-create.facade';

export const TodoCreateContainer = memo(() => {
  const { isFetching, create } = useTodoCreateFacade();

  return <TodoCreate isFetching={isFetching} onCreate={create} />;
});
