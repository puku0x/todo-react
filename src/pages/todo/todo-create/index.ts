import { lazy } from 'react';

export const TodoCreatePage = lazy(() =>
  import('./todo-create.page').then((m) => ({ default: m.TodoCreatePage }))
);
