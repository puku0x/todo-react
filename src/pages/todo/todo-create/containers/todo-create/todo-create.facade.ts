import { unwrapResult } from '@reduxjs/toolkit';
import { useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory } from 'react-router-dom';

import { TodoCreateDto } from '../../../../../models';
import { createTodo, isFetchingSelector } from '../../../../../store/todo';

export const useTodoCreateFacade = () => {
  const history = useHistory();
  const dispatch = useDispatch();
  const isFetching = useSelector(isFetchingSelector);

  const create = useCallback(
    (dto: TodoCreateDto) => {
      return dispatch(createTodo({ todo: dto }))
        .then(unwrapResult)
        .then((payload) => {
          const { todo } = payload;
          history.push(`/todos/${todo.id}`);
        });
    },
    [history, dispatch]
  );

  return {
    isFetching,
    create,
  } as const;
};
