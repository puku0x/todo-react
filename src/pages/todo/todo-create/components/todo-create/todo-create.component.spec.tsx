import { act, render } from '@testing-library/react';
import * as React from 'react';
import { MemoryRouter } from 'react-router-dom';

import { TodoCreate } from './todo-create.component';

describe('TodoCreate', () => {
  it('render', async () => {
    let result = render(<React.Fragment />);
    await act(async () => {
      result = render(<TodoCreate />, {
        wrapper: MemoryRouter,
      });
    });
    const { asFragment } = result;

    expect(asFragment()).toMatchSnapshot();
  });

  it('render with isFetching', async () => {
    let result = render(<React.Fragment />);
    await act(async () => {
      result = render(<TodoCreate isFetching />, {
        wrapper: MemoryRouter,
      });
    });
    const { asFragment } = result;

    expect(asFragment()).toMatchSnapshot();
  });
});
