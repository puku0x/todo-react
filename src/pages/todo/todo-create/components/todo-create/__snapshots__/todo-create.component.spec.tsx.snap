// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoCreate render 1`] = `
<DocumentFragment>
  <a
    href="/todos"
  >
    Back to list
  </a>
  <h2>
    todo-create
  </h2>
  <form>
    <p>
      <button
        disabled=""
        type="submit"
      >
        Save
      </button>
    </p>
    <table>
      <tbody>
        <tr>
          <td>
            title
          </td>
          <td>
            <input
              name="title"
              type="text"
            />
          </td>
        </tr>
      </tbody>
    </table>
  </form>
</DocumentFragment>
`;

exports[`TodoCreate render with isFetching 1`] = `
<DocumentFragment>
  <a
    href="/todos"
  >
    Back to list
  </a>
  <h2>
    todo-create
  </h2>
  <form>
    <p>
      <button
        disabled=""
        type="submit"
      >
        Save
      </button>
    </p>
    <table>
      <tbody>
        <tr>
          <td>
            title
          </td>
          <td>
            <input
              name="title"
              type="text"
            />
          </td>
        </tr>
      </tbody>
    </table>
  </form>
</DocumentFragment>
`;
